frontend/src/utils/menus.js

```diff
@@ -1,45 +1,53 @@
 import React from "react";
 import { Link } from "react-router-dom";
-import { Menu } from "antd";
 
-// renderMenuItem and matchingMenuKeys taken from chord_web
+/**
+ * @typedef {Required<import("antd").MenuProps>['items'][number]} MenuItem
+ * @typedef {Partial<{ key: string, icon: JSX.Element, text: string, url: string, onClick: () => void, children: BadMenuItem[], style: React.CSSProperties, disabled?: boolean }>} BadMenuItem
+ */
 
-// Custom menu renderer
-export const renderMenuItem = (i) => {
-	if (i['children']) {
-		return (
-			<Menu.SubMenu
-				style={i.style || {}}
-				title={
-					<span>
-						{i.icon || null}
-									<span>{i.text || null}</span>
-					</span>
-				}
-				key={i.key || ''}
-			>
-				{(i.children || []).map((ii) => renderMenuItem(ii))}
-			</Menu.SubMenu>
-		)
+/**
+ * @param {BadMenuItem} menuItem
+ * @returns {Pick<MenuItem, "key" | "label" | "onClick" | "children">}
+ */
+export function resolveBadMenuItem(menuItem) {
+	return {
+		key: menuItem.key || menuItem.url || "",
+		onClick: menuItem.onClick,
+		style: menuItem.style,
+		disabled: menuItem.disabled ?? false,
+		label: menuItem.url && !menuItem.onClick ? (
+			<Link to={menuItem.url}>
+				{menuItem.icon || null}
+				<span className="freezeman-menu-text">{menuItem.text || null}</span>
+			</Link>
+		) : (
+			<span>
+				{menuItem.icon || null}
+				<span className="freezeman-menu-text">{menuItem.text || null}</span>
+			</span>
+		),
+		children: menuItem.children ? menuItem.children.map(resolveBadMenuItem) : undefined,
 	}
-
-	return (
-		<Menu.Item key={i.key || i.url || ''} onClick={i.onClick || undefined} style={i.style || {}} disabled={i.disabled || false}>
-			{i.url && !i.onClick ? (
-				<Link to={i.url}>
-					{i.icon || null}
-					<span className="freezeman-menu-text">{i.text || null}</span>
-				</Link>
-			) : (
-				<span>
-					{i.icon || null}
-					<span className="freezeman-menu-text">{i.text || null}</span>
-				</span>
-			)}
-		</Menu.Item>
-	)
 }
 
-export const matchingMenuKeys = menuItems => menuItems
-  .filter(i => (i.url && window.location.pathname.startsWith(i.url)) || (i.children || []).length > 0)
-  .flatMap(i => [i.key || i.url || "", ...matchingMenuKeys(i.children || [])]);
+
+/**
+* @param {BadMenuItem[]} menuItems
+* @returns {string[]}
+ */
+export function matchingMenuKeys(menuItems) {
+	/**
+	 * @type {string[]}
+	 */
+	const keys = []
+	for (const i of menuItems) {
+		if (i.url && window.location.pathname.startsWith(i.url)) {
+			keys.push(i.key || i.url || "")
+		}
+		if (i.children) {
+			keys.push(...matchingMenuKeys(i.children))
+		}
+	}
+	return keys
+}
```

Comment:
i know this file was already using the window.location since its js but this could be converted to ts (which is something you are usually keeping an eye out for).

if you convert the file, you can use the following package to get the current url values required and, as you may already know, they can be deconstructed as follow:

`import { Route, Link, Routes, useLocation } from 'react-router-dom';
const location = useLocation();
const { hash, pathname, search } = location;
`

Reply:
I'm hesitant to make any big changes. The name `useLocation` implies that it's a hook, so `matchingMenuKeys` would need to be converted to a hook which leads to more code changes.

To make matters worse, there's recursion. Not sure if hooks behave well with recursion.

Maybe it can be avoided. But again, I'm trying to reduce changes and surprises while removing warnings.

Reply:
no worries at all, these are simply observations and not mandatory changes, in my experience pr comments are mostly suggestion unless there are some major breaking changes that would require, usually common sense is to be applied instead of the usual pragmatic approach on these smaller issues